service/storage_proxy.cc

```diff
@@ -159,6 +165,26 @@ class storage_proxy::remote {
     netw::connection_drop_slot_t _connection_dropped;
     netw::connection_drop_registration_t _condrop_registration;
 
+    future<std::optional<replica::stale_topology_exception>> try_apply_fence(fencing_token fence,
+        lowres_clock::time_point timeout)
+    {
+        const auto current_version = _sp._shared_token_metadata.get()->get_topology_version();
+        if (fence.topology_version > current_version) {
+            throw std::runtime_error(format("fencing_token topology_version {} is greater "
+                                            "than the current topology version {}",
+                fence.topology_version, current_version));
+        }
+
+        if (fence.topology_version == current_version ||
+            (fence.allow_previous && (fence.topology_version + 1 == current_version)))
+        {
+            co_return std::nullopt;
+        }
+        slogger.trace("Stale topology fenced, caller version {}, callee version {}",
+            fence.topology_version, current_version);
+        co_return replica::stale_topology_exception(fence.topology_version, current_version);
```

Comment:
It's a waste to use a coroutine here. It adds an unconditional allocation, and you don't even need to wait.

We need to consider whether to return exceptions-as-values (see utils/result.hh). I keep harping about how we don't want exceptions in the normal case, but we also don't want the abnormal case to cause an exception storm. Since this has an impact on IDL, we have to make a decision beforehand.